src/shm_sub.c

```diff
@@ -1687,6 +1899,183 @@ sr_shmsub_oper_notify(const struct lys_module *ly_mod, const char *xpath, const
+        }
+        notify_subs[i].sub_shm = (sr_sub_shm_t *)notify_subs[i].shm_sub.addr;
+
+        /* SUB WRITE LOCK */
+        if ((err_info = sr_shmsub_notify_new_wrlock(notify_subs[i].sub_shm, mod->ly_mod->name, 0, conn->cid))) {
+            goto cleanup;
+        }
+        notify_subs[i].locked = 1;
+
+        /* open sub data SHM */
+        if ((err_info = sr_shmsub_data_open_remap(mod->ly_mod->name, "oper",
+                sr_str_hash(xpath, notify_subs[i].xpath_sub->priority), &notify_subs[i].shm_data_sub, 0))) {
+            goto cleanup;
+        }
+
+        /* write the request for state data */
+        request_id = notify_subs[i].sub_shm->request_id + 1;
+        if ((err_info = sr_shmsub_notify_write_event(notify_subs[i].sub_shm, conn->cid, request_id, SR_SUB_EV_OPER,
+                orig_name, orig_data, &notify_subs[i].shm_data_sub, request_xpath, parent_lyb, parent_lyb_len,
+                xpath))) {
+            goto cleanup;
+        }
+
+        /* notify using event pipe */
+        if ((err_info = sr_shmsub_notify_evpipe(notify_subs[i].xpath_sub->evpipe_num))) {
+            goto cleanup;
+        }
+    }
+
+    /* wait until the event is processed */
+    if ((err_info = sr_shmsub_notify_all_wait_wr(notify_subs, notify_count, SR_SUB_EV_ERROR, 1, timeout_ms, conn->cid, cb_err_info))) {
+        goto cleanup;
+    }
+
+    if (*cb_err_info) {
+        /* failed callback */
+        SR_LOG_WRN("Event \"operational\" with ID %" PRIu32 " failed (%s).", request_id,
+                sr_strerror((*cb_err_info)->err[0].err_code));
+        goto cleanup;
+    } else {
+        SR_LOG_INF("Event \"operational\" with ID %" PRIu32 " succeeded.", request_id);
+    }
+
+    for (i = 0; i < notify_count; ++i) {
+        assert(notify_subs[i].sub_shm->event == SR_SUB_EV_SUCCESS);
+
+        /* parse returned data */
+        if (lyd_parse_data_mem(mod->ly_mod->ctx, notify_subs[i].shm_data_sub.addr, LYD_LYB, LYD_PARSE_ONLY | LYD_PARSE_STRICT, 0,
+                &oper_data)) {
+            sr_errinfo_new_ly(&err_info, mod->ly_mod->ctx, NULL);
+            sr_errinfo_new(&err_info, SR_ERR_VALIDATION_FAILED, "Failed to parse returned \"operational\" data.");
+            goto cleanup;
+        }
+
+        /* event processed */
+        notify_subs[i].sub_shm->event = SR_SUB_EV_NONE;
+
+        /* SUB WRITE UNLOCK */
+        sr_rwunlock(&notify_subs[i].sub_shm->lock, 0, SR_LOCK_WRITE, conn->cid, __func__);
+        notify_subs[i].locked = 0;
+
+        /* merge returned data into data tree */
+        if (lyd_merge_siblings(data, oper_data, LYD_MERGE_DESTRUCT | LYD_MERGE_WITH_FLAGS)) {
+            sr_errinfo_new_ly(&err_info, mod->ly_mod->ctx, NULL);
+            goto cleanup;
+        }
+    }
+
+cleanup:
+    /* SUB WRITE UNLOCK */
+    for (i = 0; i < notify_count; ++i) {
+        if (notify_subs[i].locked) {
+            sr_rwunlock(&notify_subs[i].sub_shm->lock, 0, SR_LOCK_WRITE, conn->cid, __func__);
```

Comment:
For consistency, clear the `locked` flag.

Reply:
Done